jerry-main/cli.c

```diff
@@ -0,0 +1,429 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include "cli.h"
+
+/*
+ * Fixed layout settings
+ */
+
+/**
+ * Wrap lines at:
+ */
+#define CLI_LINE_LENGTH 80
+
+/**
+ * Indent various lines with:
+ */
+#define CLI_LINE_INDENT 2
+
+/**
+ * Tab stop (for multi-column display) at:
+ */
+#define CLI_LINE_TAB 24
+
+/*
+ * Print helper functions
+ */
+
+/**
+ * Pad with spaces.
+ */
+static void
+cli_print_pad (int cnt) /**< number of spaces to print */
+{
+  for (int i = 0; i < cnt; i++)
+  {
+    printf (" ");
+  }
+} /* cli_print_pad */
+
+/**
+ * Print the prefix of a string.
+ */
+static void
+cli_print_prefix (const char *str, /**< string to print */
+                  int len) /**< length of the prefix to print */
+{
+  for (int i = 0; i < len; i++)
+  {
+    printf ("%c", *str++);
+  }
+} /* cli_print_prefix */
+
+/**
+ * Print a help message wrapped into the second column.
+ */
+static void
+cli_print_help (const char *help) /**< the help message to print */
+{
+  for (; help != NULL && *help != 0;)
+  {
+    int length = -1;
+    int i = 0;
+    for (; i < CLI_LINE_LENGTH - CLI_LINE_TAB && help[i] != 0; i++)
+    {
+      if (help[i] == ' ')
+      {
+        length = i;
+      }
+    }
+    if (length < 0 || i < CLI_LINE_LENGTH - CLI_LINE_TAB)
+    {
+      length = i;
+    }
+
+    cli_print_prefix (help, length);
+
+    help += length;
+    while (*help == ' ' && *help != 0)
+    {
+      help++;
+    }
+
+    if (*help != 0)
+    {
+      printf ("\n");
+      cli_print_pad (CLI_LINE_TAB);
+    }
+  }
+} /* cli_print_help */
+
+/*
+ * Command line option handling
+ */
+
+/**
+ * Initialize a command line option processor.
+ *
+ * @return the state that should be passed iteratively to cli_opt_process.
+ */
+cli_opt_state_t
+cli_opt_init (const cli_opt_t *opts, /**< array of option definitions, terminated by CLI_OPT_END */
+              int argc, /**< number of command line arguments */
+              char **argv) /**< array of command line arguments */
+{
+  return (cli_opt_state_t)
```

Comment:
On many platforms, this is exactly what happens when returning structs. AFAIK both on ARM32 and X86, struct returns are done by passing a hidden pointer argument to the function, where it has to write the results. So, `x = f();` and `f(&x);` are exactly the same. (The former might work better with optimizations, as the ABI-enforced extra pointer argument might be easier to track in pointer aliasing algorithms than completely generic pointer arguments, but that can be use case specific, whether there is a real difference).

(BTW, this is not performance-critical code (even if there would be any performance penalty). This function is typically called only once in an application. And this is for jerry-main only, not for anyone to use.)